File: src/lib/watches.ts
import { invoke } from "@tauri-apps/api/core";
import { getSeriesGroupKey, parseSeasonFromTitle, stripSeasonSuffix } from "@/lib/episodeName";

export interface Watch {
  id: number;
  title: string;
  query: string;
  anilist_id: number | null;
  cover_url: string | null;
  quality: string;
  audio_lang: string | null;
  sub_lang: string | null;
  folder: string;
  delete_after_days: number | null;
  rating: number | null;
  active: boolean;
  notify_on_available: boolean;
  status: string | null;
  list_status: string;
  episodes: number | null;
  episode_start: number | null;
  episode_end: number | null;
  created_at: string;
  updated_at: string;
  /** "Reduzir resolução" desse anime: "720p" = ligado; null/"original" = desligado.
   * Ligado na Config vale pra todos por cima disso. */
  max_resolution: string | null;
  /** "Remover áudios extras" desse anime — mesma regra. */
  strip_audio: boolean;
  /** Anime (1ª temporada na AniList) dessa temporada — agrupa na Biblioteca.
   * null = ainda não resolvido (cai no agrupamento por título). */
  series_anilist_id: number | null;
  series_title: string | null;
}

/** Chave de agrupamento por anime: franquia da AniList quando resolvida,
 * senão o título sem "Season N". */
export function seriesKeyOf(w: Watch): string {
  return w.series_anilist_id != null ? `anilist:${w.series_anilist_id}` : `title:${getSeriesGroupKey(w.title)}`;
}

export function seriesTitleOf(w: Watch): string {
  return w.series_title ?? stripSeasonSuffix(w.title);
}

export interface WatchGroup {
  key: string;
  title: string;
  representative: Watch;
  seasons: Watch[];
  /** Atualização mais recente entre as temporadas. */
  updated_at: string;
}

// Temporadas do mesmo anime viram UM card só na grade (título do anime,
// sem "Season N") — a temporada vira aba lá dentro. Agrupa pela franquia na
// AniList (pega temporada sem "Season N" no nome, ex. "Entertainment
// District Arc"), ou pelo título quando ainda não resolvida.
// Representante = temporada mais recente (maior anilist_id ≈ mais nova).
export function groupBySeries(watches: Watch[]): WatchGroup[] {
  const map = new Map<string, Watch[]>();
  for (const w of watches) {
    const key = seriesKeyOf(w);
    const arr = map.get(key);
    if (arr) arr.push(w);
    else map.set(key, [w]);
  }
  return Array.from(map.entries()).map(([key, seasons]) => {
    const sorted = [...seasons].sort(
      (a, b) =>
        (parseSeasonFromTitle(b.title) ?? 0) - (parseSeasonFromTitle(a.title) ?? 0) ||
        (b.anilist_id ?? 0) - (a.anilist_id ?? 0),
    );
    const updated_at = seasons.map((s) => s.updated_at).reduce((a, b) => (new Date(b) > new Date(a) ? b : a));
    return { key, title: seriesTitleOf(sorted[0]), representative: sorted[0], seasons: sorted, updated_at };
  });
}

export type ListStatus = "watching" | "downloaded" | "completed" | "planning";

export interface NewWatch {
  title: string;
  query: string;
  anilist_id?: number | null;
  cover_url?: string | null;
  quality?: string | null;
  audio_lang?: string | null;
  sub_lang?: string | null;
  folder?: string | null;
  delete_after_days?: number | null;
  active?: boolean | null;
  notify_on_available?: boolean | null;
  status?: string | null;
  list_status?: string | null;
  episodes?: number | null;
  episode_start?: number | null;
  episode_end?: number | null;
  max_resolution?: string | null;
  strip_audio?: boolean | null;
}

export async function listWatches(): Promise<Watch[]> {
  return invoke<Watch[]>("list_watches");
}

export async function createWatch(watch: NewWatch): Promise<Watch> {
  return invoke<Watch>("create_watch", { watch });
}

export async function deleteWatch(id: number): Promise<void> {
  return invoke<void>("delete_watch", { id });
}

export async function setWatchActive(id: number, active: boolean): Promise<void> {
  return invoke<void>("set_watch_active", { id, active });
}

export async function setWatchRating(id: number, rating: number | null): Promise<void> {
  return invoke<void>("set_watch_rating", { id, rating });
}

export async function setWatchListStatus(id: number, listStatus: ListStatus): Promise<void> {
  return invoke<void>("set_watch_list_status", { id, listStatus });
}

export interface WatchPreferences {
  quality: string;
  audio_lang: string | null;
  sub_lang: string | null;
  delete_after_days: number | null;
  notify_on_available: boolean;
  episode_start: number | null;
  episode_end: number | null;
  max_resolution: string | null;
  strip_audio: boolean;
}

export async function setWatchPreferences(id: number, prefs: WatchPreferences): Promise<void> {
  return invoke<void>("set_watch_preferences", { id, prefs });
}

File: src-tauri/src/db/watches.rs
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, SqlitePool};

#[derive(Debug, Serialize, FromRow)]
pub struct Watch {
    pub id: i64,
    pub title: String,
    pub query: String,
    pub anilist_id: Option<i64>,
    /// ID no MyAnimeList — resolvido sob demanda (não no create) via campo
    /// `idMal` da AniList, na 1ª vez que o player precisa buscar skip
    /// segments (ver `sources::aniskip`). `None` até então.
    pub mal_id: Option<i64>,
    pub cover_url: Option<String>,
    pub quality: String,
    pub audio_lang: Option<String>,
    pub sub_lang: Option<String>,
    pub folder: String,
    pub delete_after_days: Option<i64>,
    pub rating: Option<i64>,
    pub active: bool,
    pub notify_on_available: bool,
    pub status: Option<String>,
    pub list_status: String,
    pub episodes: Option<i64>,
    pub episode_start: Option<i64>,
    pub episode_end: Option<i64>,
    pub created_at: String,
    pub updated_at: String,
    /// "Reduzir resolução" desse anime: "720p" = ligado; NULL/"original" =
    /// desligado. A opção global em Config, quando ligada, vale por cima.
    pub max_resolution: Option<String>,
    /// "Remover áudios extras" desse anime — mesma regra da global.
    pub strip_audio: bool,
    /// Anime (1ª temporada da franquia na AniList) a que essa temporada
    /// pertence — chave de agrupamento na Biblioteca. NULL = ainda não
    /// resolvido (sem anilist_id, ou AniList fora do ar).
    pub series_anilist_id: Option<i64>,
    pub series_title: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct NewWatch {
    pub title: String,
    pub query: String,
    pub anilist_id: Option<i64>,
    pub cover_url: Option<String>,
    pub quality: Option<String>,
    pub audio_lang: Option<String>,
    pub sub_lang: Option<String>,
    pub folder: Option<String>,
    pub delete_after_days: Option<i64>,
    pub active: Option<bool>,
    pub notify_on_available: Option<bool>,
    pub status: Option<String>,
    pub list_status: Option<String>,
    pub episodes: Option<i64>,
    pub episode_start: Option<i64>,
    pub episode_end: Option<i64>,
    pub max_resolution: Option<String>,
    pub strip_audio: Option<bool>,
    /// Preenchidos pelo backend (ver `commands::watches::create_watch`),
    /// não pelo front.
    #[serde(skip)]
    pub series_anilist_id: Option<i64>,
    #[serde(skip)]
    pub series_title: Option<String>,
}

fn sanitize_folder_name(name: &str) -> String {
    name.chars().filter(|c| !"<>:\"/\\|?*".contains(*c)).collect()
}

pub async fn list(pool: &SqlitePool) -> Result<Vec<Watch>, sqlx::Error> {
    sqlx::query_as::<_, Watch>("SELECT * FROM watches ORDER BY created_at DESC")
        .fetch_all(pool)
        .await
}

pub async fn get(pool: &SqlitePool, id: i64) -> Result<Watch, sqlx::Error> {
    sqlx::query_as::<_, Watch>("SELECT * FROM watches WHERE id = ?")
        .bind(id)
        .fetch_one(pool)
        .await
}

pub async fn create(pool: &SqlitePool, library_root: &str, w: NewWatch) -> Result<Watch, sqlx::Error> {
    let now = chrono::Utc::now().to_rfc3339();
    let folder = w.folder.unwrap_or_else(|| {
        // Pasta = nome do ANIME, não da temporada: o Jellyfin só agrupa como
        // a mesma série se todas as temporadas morarem na mesma pasta pai.
        // Nome do anime vem da 1ª temporada na AniList quando resolvido
        // (cobre temporada sem "Season N" no título); senão, tira o
        // "Season N" do título.
        let base_title = match &w.series_title {
            Some(series) => series.clone(),
            None => crate::sources::nyaa::split_season(&w.title).0,
        };
        // `Path::join` usa o separador certo do SO — um `format!("{}/{}", ...)`
        // aqui misturava "\" (do library_root do Windows) com "/" (literal),
        // e esse path quebrado ia parar em `episodes.save_path`, fazendo o
        // botão "Abrir pasta" falhar silenciosamente (bug real reportado).
        std::path::Path::new(library_root)
            .join(sanitize_folder_name(&base_title))
            .to_string_lossy()
            .to_string()
    });
    let quality = w.quality.unwrap_or_else(|| "1080p".to_string());
    let active = w.active.unwrap_or(true);
    let notify_on_available = w.notify_on_available.unwrap_or(true);
    let list_status = w.list_status.unwrap_or_else(|| "watching".to_string());

    let id: i64 = sqlx::query_scalar(
        "INSERT INTO watches \
         (title, query, anilist_id, cover_url, quality, audio_lang, sub_lang, folder, delete_after_days, active, notify_on_available, status, list_status, episodes, episode_start, episode_end, max_resolution, strip_audio, series_anilist_id, series_title, created_at, updated_at) \
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) RETURNING id",
    )
    .bind(&w.title)
    .bind(&w.query)
    .bind(w.anilist_id)
    .bind(&w.cover_url)
    .bind(&quality)
    .bind(&w.audio_lang)
    .bind(&w.sub_lang)
    .bind(&folder)
    .bind(w.delete_after_days)
    .bind(active)
    .bind(notify_on_available)
    .bind(&w.status)
    .bind(&list_status)
    .bind(w.episodes)
    .bind(w.episode_start)
    .bind(w.episode_end)
    .bind(&w.max_resolution)
    .bind(w.strip_audio.unwrap_or(false))
    .bind(w.series_anilist_id)
    .bind(&w.series_title)
    .bind(&now)
    .bind(&now)
    .fetch_one(pool)
    .await?;

    get(pool, id).await
}

pub async fn set_series(pool: &SqlitePool, id: i64, series_anilist_id: i64, series_title: &str) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE watches SET series_anilist_id = ?, series_title = ? WHERE id = ?")
        .bind(series_anilist_id)
        .bind(series_title)
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn set_mal_id(pool: &SqlitePool, id: i64, mal_id: i64) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE watches SET mal_id = ? WHERE id = ?")
        .bind(mal_id)
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn delete(pool: &SqlitePool, id: i64) -> Result<(), sqlx::Error> {
    sqlx::query("DELETE FROM watches WHERE id = ?").bind(id).execute(pool).await?;
    Ok(())
}

pub async fn set_active(pool: &SqlitePool, id: i64, active: bool) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE watches SET active = ?, updated_at = ? WHERE id = ?")
        .bind(active)
        .bind(chrono::Utc::now().to_rfc3339())
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn set_rating(pool: &SqlitePool, id: i64, rating: Option<i64>) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE watches SET rating = ?, updated_at = ? WHERE id = ?")
        .bind(rating)
        .bind(chrono::Utc::now().to_rfc3339())
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn set_list_status(pool: &SqlitePool, id: i64, list_status: &str) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE watches SET list_status = ?, updated_at = ? WHERE id = ?")
        .bind(list_status)
        .bind(chrono::Utc::now().to_rfc3339())
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

#[derive(Debug, Deserialize)]
pub struct WatchPreferences {
    pub quality: String,
    pub audio_lang: Option<String>,
    pub sub_lang: Option<String>,
    pub delete_after_days: Option<i64>,
    pub notify_on_available: bool,
    pub episode_start: Option<i64>,
    pub episode_end: Option<i64>,
    #[serde(default)]
    pub max_resolution: Option<String>,
    #[serde(default)]
    pub strip_audio: bool,
}

pub async fn set_preferences(
    pool: &SqlitePool,
    id: i64,
    prefs: WatchPreferences,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        "UPDATE watches SET quality = ?, audio_lang = ?, sub_lang = ?, delete_after_days = ?, \
         notify_on_available = ?, episode_start = ?, episode_end = ?, max_resolution = ?, \
         strip_audio = ?, updated_at = ? WHERE id = ?",
    )
    .bind(&prefs.quality)
    .bind(&prefs.audio_lang)
    .bind(&prefs.sub_lang)
    .bind(prefs.delete_after_days)
    .bind(prefs.notify_on_available)
    .bind(prefs.episode_start)
    .bind(prefs.episode_end)
    .bind(&prefs.max_resolution)
    .bind(prefs.strip_audio)
    .bind(chrono::Utc::now().to_rfc3339())
    .bind(id)
    .execute(pool)
    .await?;
    Ok(())
}
